extra/src/main/java/aima/extra/probability/BigDecimalProbabilityNumber.java

```diff
@@ -0,0 +1,381 @@
+	}
+
+	/**
+	 * Checks if argument implementing ProbabilityNumber interface is equal to
+	 * the value of the current BigDecimalProbabilityNumber. The check between
+	 * two ProbabilityNumbers involves a check only upto the minimum precision
+	 * of the two values, i.e min(this.precision(), that.precision()).
+	 * 
+	 * @param that
+	 *            the ProbabilityNumber type that is to be compared with this
+	 *            BigDecimalProbabilityNumber.
+	 *
+	 * @return true if this == that upto precision value specified by
+	 *         min(this.precision(), that.precision()), false otherwise.
+	 */
+	@Override
+	public boolean equals(ProbabilityNumber that) {
+		BigDecimalProbabilityNumber specificType = toInternalType(that);
+		return (compareBigDecimal(this.value, this.precisionMathContext, specificType.getValue(), specificType.getMathContext()) == 0);
+	}
+	
+	/**
+	 * Compare this with another ProbabilityNumber value (that).
+	 * 
+	 * @param that
+	 *            of type ProbabilityNumber.
+	 * 
+	 * @return 1 if this > that, 0 if this == that, -1 if this < that.
+	 */
+	@Override
+	public int compareTo(ProbabilityNumber that) {
+		BigDecimalProbabilityNumber second = toInternalType(that);
+		return compareBigDecimal(this.value, this.precisionMathContext, second.getValue(), second.getMathContext());
+	}
+
+	/**
+	 * Add a BigDecimalProbabilityNumber to this BigDecimalProbabilityNumber and
+	 * return a new BigDecimalProbabilityNumber.
+	 * 
+	 * @param that
+	 *            the BigDecimalProbabilityNumber that is to be added to this
+	 *            BigDecimalProbabilityNumber.
+	 * 
+	 * @return a new BigDecimalProbabilityNumber that is the result of addition.
+	 */
+	@Override
+	public ProbabilityNumber add(ProbabilityNumber that) {
+		BigDecimalProbabilityNumber addend = toInternalType(that);
+		int minPrecision = getMinPrecision(this.precisionMathContext.getPrecision(), addend.getMathContext().getPrecision());
+		MathContext resultMathContext = new MathContext(minPrecision, DEFAULT_PRECISION_ROUNDING_MODE);
+		return new BigDecimalProbabilityNumber(value.add(addend.getValue(), resultMathContext), resultMathContext);
+	}
+
+	/**
+	 * Subtract a BigDecimalProbabilityNumber from this
+	 * BigDecimalProbabilityNumber and return a new BigDecimalProbabilityNumber.
+	 * 
+	 * @param that
+	 *            the BigDecimalProbabilityNumber that is to be subtracted from
+	 *            this BigDecimalProbabilityNumber.
+	 * 
+	 * @return a new BigDecimalProbabilityNumber that is the result of
+	 *         subtraction.
+	 */
+	@Override
+	public ProbabilityNumber subtract(ProbabilityNumber that) {
+		BigDecimalProbabilityNumber subtrahend = toInternalType(that);
+		int minPrecision = getMinPrecision(this.precisionMathContext.getPrecision(),
+				subtrahend.getMathContext().getPrecision());
+		MathContext resultMathContext = new MathContext(minPrecision, DEFAULT_PRECISION_ROUNDING_MODE);
```

Comment:
I think another method getResultMathContext() should be written to reduce the repeated code above.